### ISSUE TITLE
Many MongoDB connections

### ISSUE THREAD

**Issue by houshuang** [share/sharedb-mongo#80]
We are using MongoDB Atlas to host Mongo, and we've been doing some load-testing recently to figure out how to properly provision our ShareDB cluster (we're currently aiming to support at least 1200, possibly 1800 concurrent users). 

While we were doing load-testing, we noticed that the number of connections to MongoDB grew at a much higher rate than we had expected. My understanding was that each ShareDB server maintains a single connection to MongoDB (unconnected to how many documents are subscribed to, etc)? We had between 10 and 20 ShareDB servers running, and at one point we had more than 200 connections to the MongoDB, close to hitting the limits for our Mongo service.

The charts attached seem to show a growing number of connections for each "wave" of ops etc. We also saw some Mongo failures, one hypothesis I have is that we overwhelmed Mongo, and that the failed connections were not properly terminated, while new ones were created. However, I'd love some insights from the creators of this library as to what might be happening and how we can avoid it. (Note that this was a completely unrealistic load, and so far in production, everything has been fine. BUt we'd still like to know what's happening in more detail).

thanks!!!

![image](https://user-images.githubusercontent.com/61575/61465939-52fd9c00-a979-11e9-81a0-9621ba67081f.png)

**Comment by alecgibson** [comment]
How are you initialising ShareDB? And how are you connecting to clients? This potentially sounds symptomatic of spinning up a new ShareDB instance and/or new Mongo connection every time a client connects...?

**Comment by houshuang** [comment]
Thanks a lot for your quick reply! That's a good intuition, but I don't think that's what's happening. Here's where we initialize ShareDB:

```js
import ShareDB from '@teamwork/sharedb';
import RedisPubsub from '@teamwork/sharedb-redis-pubsub';
import richText from '@minervaproject/rich-text';
import json from '@minervaproject/ot-json0';
import mongo from '@teamwork/sharedb-mongo';

import { SHAREDB_COLLECTION, MONGODB_URL, REDIS_URL } from './config';

const options = {
  db: MONGODB_URL ? mongo(MONGODB_URL) : null,
  pubsub: REDIS_URL ? new RedisPubsub(REDIS_URL) : null,
  disableDocAction: true,
  disableSpaceDelimitedActions: true,
};

json.type.registerSubtype(richText.type);
ShareDB.types.register(json.type);
ShareDB.types.defaultType = json.type;

const share = new ShareDB(options);

export const connection = share.connect();

export const getShareDocument = (uuid) => {
  const doc = connection.get(SHAREDB_COLLECTION, uuid);
  return new Promise((resolve, reject) => {
    doc.fetch((err) => {
      if (err) {
        reject(err);
      }
      resolve(doc);
    });
  });
};

export default share;
```

And this file should just be imported once by the startup-script... But to be completely sure, we could of course put a console.log and see how often that gets called...

**Comment by alecgibson** [comment]
Starting with a `console.log` statement sounds like a good start to me. Also I've noticed you're on a fork for _all_ of the packages; does this happen with a "vanilla" ShareDB setup? Can you tie the new Mongo connections with any ShareDB-related events by matching up timestamps?

I'm afraid you'll have to do some digging in your logs or produce a minimal working example for any further assistance.

**Comment by houshuang** [comment]
Hi,
thanks again for taking the time to answer, even though we're on forked packages :) The reason we're using the @teamwork fork is to use presence/shared cursors (which we've added to ot-json0 in our own fork). We would love to get this merged upstream, and have been working with curran, but it seems like a long process. Because we're on @teamwork/sharedb I think we need to be on @teamwork/sharedb-mongo unless we fork sharedb/sharedb-mongo and change which sharedb it imports. But I don't think this is the problem.

Anyway I did some simple testing with console.log and adding lot's of sessions and documents locally, and I could not get it to open any more connections. At this point I believe the problem is simply that we completely overloaded the connections during load-testing (much more than realistic) and that we had some hung or zombie connections. It might even have something to do with our Kubernetes cluster (maybe there were hung pods, and new pods starting on top of them). I just wanted to make sure there wasn't anything around the core functioning of sharedb-mongo that I had misunderstood. As long as this doesn't happen in production (with more reasonable loads), I think we're good.

Thanks again.
